### PR TITLE
[NZPMC-013] Implement Submission Mutation

### PR DESCRIPTION
**Pull request by Kingston802**
**Issue**

Needs to calculate and set score in firebase

**Solution**

Created submitUserQuizQuestions(), setUserQuizScore()

**Discussion**
Unsure about how I implemented the update in setUserQuizScore, I tried to use the .update method but I couldn't make it work because the method didn't exist on the UserQuiz object (because it was 'packed'?). Would appreciate some review and then I'll refactor. 

Additionally, the Timestamp object is not the one in the firebase docs, it doesn't have the toDate() method but the model requires a date object hence what you see in the code. 

**Risk**

Low risk

**Reviewers**
Will, CZ

### CODE REVIEW

File: controllers/userQuiz.js [hunk 1]
```diff
@@ -57,4 +57,20 @@ const addUserQuiz = async (user, quiz, startTime, endTime) => {
     return await getUserQuiz(userQuiz.id)
 }
 
-export { addUserQuiz, getUserQuiz, getUserQuizzes, getAllUserQuizzes }
+const setUserQuizScore = async (userQuizID, score) => {
+    let obj = await getUserQuiz(userQuizID)
```
Comment (scorpionknifes) [inline controllers/userQuiz.js:61]: name it properly thx

File: controllers/userQuizQuestion.js [hunk 2]
```diff
@@ -125,11 +126,47 @@ const getUserQuizQuestionOptionsByQuestion = async (question) => {
...
+    let userAnswerIDs = []
+    
+    Promise.all(userAnswerIDsRequest).then((answerIDObjs) => {
+        answerIDObjs.forEach((obj) => {
+            userAnswerIDs.push(obj.id)
+        })
+    })
+
+    return userAnswerIDs
+}
```
Comment (scorpionknifes) [inline controllers/userQuizQuestion.js:146]: me don't like. 

also, I hate people calling stuff obj. Like either make it `answerIDObj` or just use a single letter `a` (not perfect but me take). like obj can mean anything.

prefer:
```js
// example don't use directly

const userAnswerIDs = await Promise.all(userAnswer.map(async (userAnswer) => {
	let userAnswerObj = userAnswer.userAnswerObj
   	const answer = await userAnswerObj.answer.get() 
   	
	return answer.id
}))

```

File: controllers/userQuizQuestion.js [hunk 3]
```diff
@@ -125,11 +126,47 @@ const getUserQuizQuestionOptionsByQuestion = async (question) => {
...
+
+const submitUserQuizQuestions = async (userQuizID, userAnswers, correctAnswers) => {
+    // calculate score
+    let score = 0
+    userAnswers.forEach((answer, index) => {
+        if (answer === correctAnswers[index]) {
+            score += 1
+        }
+    });
```
Comment (scorpionknifes) [inline controllers/userQuizQuestion.js:155]: try practice using `Array.reduce()`.

This is a very good case to use reduce, highly recommend u try it out cause u don't get many chances to use reduce.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce

**Review by scorpionknifes — CHANGES_REQUESTED** [review]

**Comment by scorpionknifes** [comment]
Excellent, keep up the good work.

**Review by Will-ZYS — APPROVED** [review]
Might wanna uncomment the AuthenticationErrors, otherwise, other than the changes CZ requested, LGTM

File: controllers/userQuiz.js (hunk 1 above)
Comment (Kingston802) [reply controllers/userQuiz.js:61]: Sorry, name the function again or obj?

Comment (Will-ZYS) [reply controllers/userQuiz.js:61]: I think he meant the variable name 'obj'

**Review by scorpionknifes — APPROVED** [review]
LGTM!

**Pull request merged**